### ISSUE TITLE
_dynamicReplacement could not call origin async method 

### ISSUE THREAD

**Issue by Whirlwind** [apple/swift#62214]
Xcode 14.1, swift-driver version: 1.62.15 Apple Swift version 5.7.1 (swiftlang-5.7.1.135.3 clang-1400.0.29.51)

```

dynamic
public func runCommander() async {
    printf("1")
}

@_dynamicReplacement(for: runCommander())
public func core_runCommander() async {
    await runCommander()
}
```

I get a deadloop:
<img width="1721" alt="image" src="https://user-images.githubusercontent.com/1251192/203046395-17184ac6-fb60-44e7-828a-046698ed8b2f.png">

I try other methods, same wrong.

**Comment by LucianoPAlmeida** [comment]
This looks like correct behavior, by telling that `core_runCommander` is a dynamic replacement for `runCommander` calling `runCommander` means calling `core_runCommander` which make this function a recursive call.

**Comment by Whirlwind** [comment]
> This looks like correct behavior, by telling that `core_runCommander` is a dynamic replacement for `runCommander` calling `runCommander` means calling `core_runCommander` which make this function a recursive call.

I don't think so. If I remove the `async`, call the `runCommander` will call the origin method:
```
dynamic
public func runCommander() {
    printf("1")
}

@_dynamicReplacement(for: runCommander())
public func core_runCommander() {
    runCommander()
}
```
It will print the `1`.

If you are right, what should I do when calling the origin method?

I found a test: https://github.com/apple/swift/blob/7123d2614b5f222d03b3762cb110d27a9dd98e24/test/Interpreter/Inputs/dynamic_replacement_chaining_B.swift
It will call the origin method by origin name.

**Comment by LucianoPAlmeida** [comment]
Ok, looks like I misunderstood the feature behavior

**Comment by LucianoPAlmeida** [comment]
Yeah this looks indeed a bug

**Comment by Whirlwind** [comment]
Any process?

**Comment by LucianoPAlmeida** [comment]
cc @aschwaighofer

**Comment by Whirlwind** [comment]
Any process?

**Comment by pbartolome** [comment]
Still reproducible with swift 5.9 `swift-driver version: 1.87.1 Apple Swift version 5.9 (swiftlang-5.9.0.128.108 clang-1500.0.40.1)`

The following sample code:

```swift
dynamic func original() {
    print("original")
}

@_dynamicReplacement(for: original)
func replacement() {
    print("replacement")
    original()
}

original()
```
correctly prints: 
```
replacement
original
```


But adding `async` to the `original` method forces the use of `await` in the replacement in order to call it, this produces a call to the replacement instead of the original implementation, ending with an infinte loop

```swift
dynamic func original() async {
    print("original")
}

@_dynamicReplacement(for: original)
func replacement() async {
    print("replacement")
    await original()
}

await original()
```
```
replacement
replacement
replacement
replacement
replacement
replacement
replacement
replacement
replacement
replacement
...
```

**Comment by maximkrouk** [comment]
Also reproducible for non-async objc methods: https://github.com/apple/swift/issues/53916

**Comment by Whirlwind** [comment]
Any process?

**Comment by Whirlwind** [comment]
Still reproducible with  Apple Swift version 6.0 (swiftlang-6.0.0.7.6 clang-1600.0.24.1)
Target: arm64-apple-macosx15.0
